Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.h

```diff
@@ -0,0 +1,105 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/ByteString.h>
+#include <AK/FlyString.h>
+#include <AK/QuickSort.h>
+#include <AK/String.h>
+#include <LibXML/DOM/Node.h>
+
+namespace JSSpecCompiler {
+
+struct LogicalLocation : RefCounted<LogicalLocation> {
+    StringView filename;
+    String section;
+    String step;
+};
+
+struct Location {
+    NonnullRefPtr<LogicalLocation> logical_location;
+    size_t line = 0, column = 0;
+
+    static Location global_scope();
+    static Location file_scope();
+    static Location from_xml_offset(XML::Offset offset);
+    static Location from_xml_offset(XML::Offset offset, NonnullRefPtr<LogicalLocation> logical_location);
+};
+
+class DiagnosticEngine {
+public:
+    DiagnosticEngine(StringView filename);
+
+    static DiagnosticEngine& instance();
+
+    template<typename Func>
+    void with(Func&& func)
+    {
+        m_instance = this;
+        func();
+        m_instance = nullptr;
+    }
+
+#define DEFINE_DIAGNOSTIC_FUNCTION(name_, level_)                                                                                 \
+    template<typename... Parameters>                                                                                              \
+    static void name_(Location const& location, AK::CheckedFormatString<Parameters...>&& fmtstr, Parameters const&... parameters) \
+    {                                                                                                                             \
+        DiagnosticEngine::instance().add_diagnostic({                                                                             \
+            .location = location,                                                                                                 \
+            .level = DiagnosticLevel::level_,                                                                                     \
+            .message = MUST(String::formatted(move(fmtstr), parameters...)),                                                      \
+        });                                                                                                                       \
+    }
+
+    DEFINE_DIAGNOSTIC_FUNCTION(note, Note)
+    DEFINE_DIAGNOSTIC_FUNCTION(warn, Warning)
+    DEFINE_DIAGNOSTIC_FUNCTION(error, Error)
+    DEFINE_DIAGNOSTIC_FUNCTION(fatal_error, FatalError)
+
+#undef DEFINE_DIAGNOSTIC_FUNCTION
+
+    bool has_fatal_errors() const;
+    void print_diagnostics();
+
+private:
+    enum class DiagnosticLevel {
+        Note,
+        Warning,
+        Error,
+        FatalError,
+    };
+
+    enum class UseColor {
+        No,
+        Yes,
+    };
+
+    struct Diagnostic {
+        Location location;
+        DiagnosticLevel level;
+        String message;
+
+        Vector<Diagnostic> notes;
+
+        bool operator<(Diagnostic const& other) const;
+
+        void format_into(StringBuilder& builder, UseColor) const;
+    };
+
+    friend Location;
```

Comment:
If the only thing we need from the DiagnosticEngine is the current file, I would prefer a public "filename()" getter to friendship.